modules/renter/hostdb/negotiate.go

```diff
@@ -418,9 +432,43 @@ func (r *Renter) newHostUploader(settings modules.HostSettings, filesize uint64,
 	if err := encoding.WriteObject(hu.conn, modules.RPCRevise); err != nil {
 		return nil, err
 	}
-	if err := encoding.WriteObject(hu.conn, hu.contract.ID); err != nil {
+	if err := encoding.WriteObject(hu.conn, hu.fcid); err != nil {
 		return nil, err
 	}
 
 	return hu, nil
 }
+
+// UniqueHosts will return up to 'n' unique hosts that are not in 'old'. Note
+// that this is a blocking call that performs network I/O.
+func (hdb *HostDB) UniqueHosts(n int, old []Uploader) (hosts []Uploader) {
+	// remove old hosts from tree
+	hdb.mu.Lock()
+	var oldEntries []*hostEntry
+	for _, host := range old {
+		node, exists := hdb.activeHosts[host.Address()]
+		if !exists {
+			continue
+		}
+		node.removeNode()
+		delete(hdb.activeHosts, host.Address()) // probably unnecessary
+		oldEntries = append(oldEntries, node.hostEntry)
+	}
+	// select random hosts from remaining set
+	randHosts := hdb.randomHosts(n)
+	// replace removed hosts
+	for _, entry := range oldEntries {
+		hdb.insertNode(entry)
+	}
+	hdb.mu.Unlock()
+
+	// negotiate contracts with each host
+	for _, host := range randHosts {
```

Comment:
at this point, contracts get re-negotiated every time? For now, that is okay. Just clarifying.